src/main/java/net/dries007/tfc/common/blocks/plant/TFCKelpTopBlock.java

```diff
@@ -0,0 +1,154 @@
+package net.dries007.tfc.common.blocks.plant;
+
+import java.util.function.Supplier;
+
+import net.minecraft.block.*;
+import net.minecraft.fluid.Fluid;
+import net.minecraft.fluid.FluidState;
+import net.minecraft.fluid.Fluids;
+import net.minecraft.item.BlockItemUseContext;
+import net.minecraft.state.StateContainer;
+import net.minecraft.util.Direction;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.util.math.shapes.ISelectionContext;
+import net.minecraft.util.math.shapes.VoxelShape;
+import net.minecraft.util.math.vector.Vector3d;
+import net.minecraft.world.IBlockReader;
+import net.minecraft.world.IWorld;
+import net.minecraft.world.World;
+
+import net.dries007.tfc.common.fluids.FluidProperty;
+import net.dries007.tfc.common.fluids.IFluidLoggable;
+import net.dries007.tfc.common.fluids.TFCFluids;
+
+public abstract class TFCKelpTopBlock extends TopPlantBlock implements IFluidLoggable
+{
+    public static TFCKelpTopBlock create(AbstractBlock.Properties properties, Supplier<? extends Block> bodyBlock, Direction direction, VoxelShape shape, FluidProperty fluid)
+    {
+        return new TFCKelpTopBlock(properties, bodyBlock, direction, shape)
+        {
+            @Override
+            public FluidProperty getFluidPropertyAbstract()
+            {
+                return fluid;
+            }
+
+            @Override
+            public Supplier<? extends Block> getBodyBlockAbstract()
+            {
+                return bodyBlock;
+            }
+        };
+    }
+
+    protected TFCKelpTopBlock(AbstractBlock.Properties properties, Supplier<? extends Block> bodyBlock, Direction direction, VoxelShape shape)
+    {
+        super(properties, bodyBlock, direction, shape);
+    }
+
+    @Override
+    public BlockState getStateForPlacement(BlockItemUseContext context)
+    {
+        World world = context.getLevel();
+        BlockState state = defaultBlockState().setValue(AGE, world.getRandom().nextInt(25));
+        FluidState fluidState = world.getFluidState(context.getClickedPos());
+        if (getFluidProperty().canContain(fluidState.getType()))
+        {
+            return state.setValue(getFluidProperty(), getFluidProperty().keyFor(fluidState.getType()));
+        }
+        return null;
+    }
+
+    //Dear Alcatraz; I'm not sure what this does
+    //I'm pretty sure this is where it converts the top to the body block so I'm putting the bit of extra handling here for that
+    //Basically the Abstract tall plant system believes that we're _not_ waterlogged (so that we can use our own fluid logging)
+    //It takes 'waterlogged' as a boolean in the constructor which I hardcode to being false in the superclass
+    //So I have to recreate their handling at the top level essentially
+    @Override
```

Comment:
I can answer specific questions about `updateShape` in discord. But generally, this updates a specific block w.r.t. a given side which has just been updated. Such as fences connecting to adjacent blocks.

The fluid ticks are odd, shouldn't be scheduling fluid ticks for fluids that don't exist at the current pos (can do `state.getFluidState().getType()` for the fluid at.)